File: python/acp-agent-sdk/src/acp_agent_sdk/agent.py
"""Agent: manifest binding + capability handlers + runtime. You never
touch transport (paved-road.md) — work arrives as Temporal activities on
this agent's task queue, and the SDK owns the polyglot contract with the
orchestrator."""

import os
from collections.abc import Awaitable, Callable
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import structlog
import yaml
from acp_protocol import validate, validation_errors
from jsonschema import Draft202012Validator
from opentelemetry import trace
from temporalio.exceptions import ApplicationError

from acp_agent_sdk.answer import AnswerBuilder
from acp_agent_sdk.context import CapabilityContext
from acp_agent_sdk.errors import CapabilityError, ErrorClass
from acp_agent_sdk.model import (
    ContextualModel,
    FakeModel,
    ModelCallContext,
    ModelClient,
    ModelResponse,
)
from acp_agent_sdk.retriever import Retriever
from acp_agent_sdk.telemetry import configure_logging

Handler = Callable[[CapabilityContext, dict[str, Any]], Awaitable[dict[str, Any]]]


@dataclass
class _CountingModel:
    """Wraps the configured model so usage lands in the StepResult."""

    inner: ModelClient
    llm_calls: int = 0
    input_tokens: int = 0
    output_tokens: int = 0
    cache_read_tokens: int = 0
    cache_write_tokens: int = 0
    # Last non-None resolved model id across completions (v0 last-write-wins).
    model: str | None = None

    async def complete(self, prompt: str, *, max_tokens: int = 1024) -> ModelResponse:
        self.llm_calls += 1
        response = await self.inner.complete(prompt, max_tokens=max_tokens)
        self.input_tokens += response.input_tokens
        self.output_tokens += response.output_tokens
        self.cache_read_tokens += response.cache_read_tokens
        self.cache_write_tokens += response.cache_write_tokens
        if response.model is not None:
            self.model = response.model
        return response


@dataclass
class Agent:
    """One agent = manifest + handlers + tool bindings + eval suite."""

    manifest: dict[str, Any]
    # Mutable: run() installs a GatewayModel when none was configured.
    model: ModelClient | None = None
    retriever: Retriever | None = None
    handlers: dict[str, Handler] = field(default_factory=dict)

    def __post_init__(self) -> None:
        configure_logging()
        self.log: structlog.stdlib.BoundLogger = structlog.get_logger(self.manifest["id"])
        # Lazy unit-test fallback: an unconfigured, unserved agent still fakes.
        self._fallback_fake: FakeModel | None = None

    @classmethod
    def from_manifest(
        cls,
        manifest_path: str | Path,
        *,
        model: ModelClient | None = None,
        retriever: Retriever | None = None,
    ) -> "Agent":
        raw = yaml.safe_load(Path(manifest_path).read_text(encoding="utf-8"))
        validate("agent-manifest", raw)
        return cls(manifest=raw, model=model, retriever=retriever)

    @property
    def agent_id(self) -> str:
        agent_id: str = self.manifest["id"]
        return agent_id

    @property
    def task_queue(self) -> str:
        return f"agent-{self.agent_id}"

    def capability(self, name: str) -> Callable[[Handler], Handler]:
        """Registers the handler for a manifest-declared capability."""
        declared = {c["name"] for c in self.manifest["capabilities"]}
        if name not in declared:
            raise ValueError(
                f"capability {name} is not declared in the manifest for {self.agent_id} — "
                f"declared: {', '.join(sorted(declared))}. Add it to manifest.yaml first; "
                "the manifest is the contract."
            )
        if name in self.handlers:
            raise ValueError(f"capability {name} already has a handler")

        def register(handler: Handler) -> Handler:
            self.handlers[name] = handler
            return handler

        return register

    def assert_complete(self) -> None:
        """Every declared capability must have a handler before serving."""
        missing = [
            c["name"] for c in self.manifest["capabilities"] if c["name"] not in self.handlers
        ]
        if missing:
            raise RuntimeError(
                f"manifest declares capabilities with no handler: {', '.join(missing)}"
            )

    async def execute(self, request: dict[str, Any]) -> dict[str, Any]:
        """The execute_capability activity body: StepRequest → StepResult.

        Output is validated against the capability's declared
        output_schema with one structured-repair retry, then the step
        fails typed — never a best-effort parse (orchestration.md).
        """
        errors = validation_errors("task-contract", request, "#/$defs/step_request")
        if errors:
            raise ApplicationError(
                f"malformed StepRequest: {'; '.join(errors)}", type="Permanent", non_retryable=True
            )
        capability = request["capability"]
        handler = self.handlers.get(capability)
        if handler is None:
            return self._failed(
                request,
                CapabilityError(
                    ErrorClass.PERMANENT,
                    f"agent {self.agent_id} has no handler for {capability}",
                ),
            )

        # A contextual model (the GatewayModel) is bound to THIS step's
        # delegated identity + correlation before it reaches the handler;
        # FakeModel is not contextual, so unit tests see zero change.
        base: ModelClient
        if self.model is not None:
            base = self.model
        else:
            if self._fallback_fake is None:
                self._fallback_fake = FakeModel()
            base = self._fallback_fake
        if isinstance(base, ContextualModel):
            base = base.with_call_context(
                ModelCallContext(
                    task_id=request["task_id"],
                    step_id=request["step_id"],
                    tenant=request["tenant"],
                    capability=capability,
                    delegated_token=request.get("delegated_token"),
                )
            )

        counting = _CountingModel(base)
        ctx = CapabilityContext(
            tenant=request["tenant"],
            task_id=request["task_id"],
            step_id=request["step_id"],
            capability=capability,
            delegated_token=request.get("delegated_token"),
            budget=request.get("budget"),
            model=counting,
            _retriever=self.retriever,
            log=self.log.bind(task_id=request["task_id"], step_id=request["step_id"]),
        )

        tracer = trace.get_tracer("acp-agent-sdk")
        with tracer.start_as_current_span(f"invoke_agent {capability}") as span:
            span.set_attributes(
                {
                    "gen_ai.operation.name": "invoke_agent",
                    "gen_ai.agent.name": self.agent_id,
                    "acp.tenant": request["tenant"],
                    "acp.task_id": request["task_id"],
                    "acp.capability": capability,
                }
            )
            try:
                output = await self._execute_validated(handler, ctx, request)
            except CapabilityError as err:
                if err.error_class is ErrorClass.RETRYABLE:
                    # Temporal owns retries; surface as a retryable failure.
                    raise ApplicationError(str(err), type="Retryable") from err
                span.set_attribute("acp.step_status", "failed")
                return self._failed(request, err, counting)
            span.set_attribute("acp.step_status", "completed")

        return {
            "kind": "step_result",
            "step_id": request["step_id"],
            "task_id": request["task_id"],
            "tenant": request["tenant"],
            "status": "completed",
            "output": output,
            "usage": self._usage(counting),
        }

    async def _execute_validated(
        self, handler: Handler, ctx: CapabilityContext, request: dict[str, Any]
    ) -> dict[str, Any]:
        declared = next(c for c in self.manifest["capabilities"] if c["name"] == ctx.capability)
        validator = Draft202012Validator(declared["output_schema"])
        output = await handler(ctx, request.get("input", {}))
        violations = [e.message for e in validator.iter_errors(output)]
        if not violations:
            return output
        ctx.log.warning(
            "output failed schema validation; structured-repair retry", errors=violations
        )
        output = await handler(ctx, request.get("input", {}))
        violations = [e.message for e in validator.iter_errors(output)]
        if violations:
            raise CapabilityError(
                ErrorClass.PERMANENT,
                f"handler output does not conform to the declared output_schema after one repair "
                f"retry: {'; '.join(violations[:3])}",
            )
        return output

    def _failed(
        self,
        request: dict[str, Any],
        err: CapabilityError,
        counting: _CountingModel | None = None,
    ) -> dict[str, Any]:
        result: dict[str, Any] = {
            "kind": "step_result",
            "step_id": request["step_id"],
            "task_id": request["task_id"],
            "tenant": request["tenant"],
            "status": "failed",
            "error": err.to_protocol(),
        }
        if counting is not None:
            result["usage"] = self._usage(counting)
        return result

    @staticmethod
    def _usage(counting: _CountingModel) -> dict[str, int | str]:
        # Cache tokens and model id are emitted only when present, keeping
        # zero-LLM usage byte-identical to before the cache fields existed and
        # matching the TypeScript SDK's usageOf().
        usage: dict[str, int | str] = {
            "llm_calls": counting.llm_calls,
            "input_tokens": counting.input_tokens,
            "output_tokens": counting.output_tokens,
        }
        if counting.cache_read_tokens > 0:
            usage["cache_read_tokens"] = counting.cache_read_tokens
        if counting.cache_write_tokens > 0:
            usage["cache_write_tokens"] = counting.cache_write_tokens
        if counting.model is not None:
            usage["model"] = counting.model
        return usage

    def answer_builder(self) -> AnswerBuilder:
        return AnswerBuilder()

    async def run(self) -> None:  # pragma: no cover — needs live Temporal+NATS; E2E covers it
        """Serves this agent's task queue until cancelled."""
        from nats.aio.client import Client as NatsClient
        from temporalio.client import Client as TemporalClient
        from temporalio.contrib.opentelemetry import TracingInterceptor
        from temporalio.worker import Worker

        from acp_agent_sdk.bus import BusTokenSource
        from acp_agent_sdk.retriever import NatsRetriever, TokenExchanger
        from acp_agent_sdk.telemetry import configure_tracing

        self.assert_complete()
        configure_tracing(self.agent_id)

        # A served agent with no configured model completes through the LLM
        # Gateway on its manifest's first allowed class (the NatsRetriever
        # precedent). Unit-tested agents keep the FakeModel fallback.
        if self.model is None:
            from acp_agent_sdk.gateway_model import GatewayModel

            allowed = (self.manifest.get("models") or {}).get("allowed") or []
            self.model = GatewayModel(
                url=os.environ.get("ACP_LLM_GATEWAY_URL", "http://localhost:7107"),
                model_class=allowed[0] if allowed else "default-tier",
            )

        if self.retriever is None:
            token_url = os.environ.get("ACP_TOKEN_URL", "http://localhost:7101")
            client_id = os.environ.get("ACP_AGENT_CLIENT_ID", f"agent-{self.agent_id}")
            client_secret = os.environ["ACP_AGENT_CLIENT_SECRET"]
            # Item 0c: the bus identity is minted from a platform JWT via the
            # auth callout — no static NATS user. A background refresh keeps
            # the token live and pushes each new one onto the connection so a
            # reconnect presents a fresh credential.
            bus_tokens = BusTokenSource(
                token_url=token_url, client_id=client_id, client_secret=client_secret
            )
            await bus_tokens.start()
            nc = NatsClient()

            def _update_token(token: str) -> None:
                nc.options["token"] = token

            bus_tokens.on_refresh = _update_token
            await nc.connect(
                servers=os.environ.get("ACP_NATS_URL", "nats://localhost:4222"),
                token=bus_tokens.token(),
            )
            self.retriever = NatsRetriever(
                nc=nc,
                exchanger=TokenExchanger(
                    token_url=token_url,
                    client_id=client_id,
                    client_secret=client_secret,
                ),
            )

        client = await TemporalClient.connect(
            os.environ.get("ACP_TEMPORAL_ADDRESS", "localhost:7233"),
            namespace=os.environ.get("ACP_TEMPORAL_NAMESPACE", "default"),
            interceptors=[TracingInterceptor()],
        )

        async def execute_capability(request: dict[str, Any]) -> dict[str, Any]:
            return await self.execute(request)

        from temporalio import activity as temporal_activity

        activity_impl = temporal_activity.defn(name="execute_capability")(execute_capability)

        worker = Worker(
            client,
            task_queue=self.task_queue,
            activities=[activity_impl],
            interceptors=[TracingInterceptor()],
        )
        self.log.info("agent worker serving", task_queue=self.task_queue)
        await worker.run()

File: packages/agent-sdk/src/agent.ts
/**
 * Agent: manifest binding + capability handlers + runtime. You never touch
 * transport (paved-road.md) — work arrives as Temporal activities on this
 * agent's task queue, and the SDK owns the polyglot contract with the
 * orchestrator.
 */

import { readFileSync } from 'node:fs';
import { trace } from '@opentelemetry/api';
import { ApplicationFailure } from '@temporalio/common';
import { Ajv2020, type ValidateFunction } from 'ajv/dist/2020.js';
import type { Logger } from 'pino';
import YAML from 'yaml';
import { agentManifest, stepRequest, type AgentManifest, type StepRequest } from '@acp/protocol';
import type { StepResult, Usage } from '@acp/protocol';
import { AnswerBuilder } from './answer.js';
import { CapabilityContext } from './context.js';
import { CapabilityError, ErrorClass } from './errors.js';
import { FakeModel, isContextualModel, type ModelClient, type ModelResponse } from './model.js';
import type { Retriever } from './retriever.js';
import { createAgentLogger } from './telemetry.js';

/** A capability handler: context + validated input → output conforming to the declared schema. */
export type Handler = (
  ctx: CapabilityContext,
  input: Record<string, unknown>,
) => Promise<Record<string, unknown>>;

export interface AgentOptions {
  manifest: AgentManifest;
  model?: ModelClient;
  retriever?: Retriever;
}

/** Generated `{}`-typed fields carry objects on the wire; narrow them once, here. */
function asRecord(value: unknown): Record<string, unknown> {
  return value as Record<string, unknown>;
}

/** Wraps the configured model so usage lands in the StepResult. */
class CountingModel implements ModelClient {
  llmCalls = 0;
  inputTokens = 0;
  outputTokens = 0;
  cacheReadTokens = 0;
  cacheWriteTokens = 0;
  /** Last non-undefined resolved model id seen across completions (v0 last-write-wins). */
  model: string | undefined;

  constructor(private readonly inner: ModelClient) {}

  async complete(prompt: string, options?: { maxTokens?: number }): Promise<ModelResponse> {
    this.llmCalls += 1;
    const response = await this.inner.complete(prompt, options);
    this.inputTokens += response.inputTokens ?? 0;
    this.outputTokens += response.outputTokens ?? 0;
    this.cacheReadTokens += response.cacheReadTokens ?? 0;
    this.cacheWriteTokens += response.cacheWriteTokens ?? 0;
    if (response.model !== undefined) this.model = response.model;
    return response;
  }
}

/** One agent = manifest + handlers + tool bindings + eval suite. */
export class Agent {
  readonly manifest: AgentManifest;
  /** Mutable: `serveAgent()` installs a GatewayModel when none was configured. */
  model: ModelClient | undefined;
  /** Mutable: `run()` installs a NatsRetriever when none was configured. */
  retriever: Retriever | undefined;
  /** Lazy unit-test fallback: an unconfigured, unserved agent still fakes. */
  private fallbackFake: FakeModel | undefined;
  readonly handlers = new Map<string, Handler>();
  readonly log: Logger;
  // The SDK's own lenient instance for capability output_schemas — authored
  // by agent teams, not protocol schemas — so their idioms don't trip Ajv
  // strict mode. Protocol documents keep using @acp/protocol's strict one.
  //
  // validateFormats: false for 1:1 parity with the Python SDK, whose
  // Draft202012Validator carries no FormatChecker — there `format:` is
  // annotation-only (jsonschema's spec-default), so it must never cause an
  // output validation failure here either. With formats off, Ajv also
  // compiles unknown format names without ajv-formats.
  private readonly ajv: Ajv2020;
  private readonly validators = new Map<string, ValidateFunction>();

  constructor(options: AgentOptions) {
    this.manifest = options.manifest;
    this.model = options.model;
    this.retriever = options.retriever;
    this.log = createAgentLogger(this.agentId);
    this.ajv = new Ajv2020({ allErrors: true, strict: false, validateFormats: false });
  }

  /** Loads + validates a manifest.yaml; throws ProtocolValidationError on drift. */
  static fromManifest(
    manifestPath: string,
    options: { model?: ModelClient; retriever?: Retriever } = {},
  ): Agent {
    const raw: unknown = YAML.parse(readFileSync(manifestPath, 'utf-8'));
    const manifest = agentManifest.parse(raw);
    const agentOptions: AgentOptions = { manifest };
    if (options.model !== undefined) agentOptions.model = options.model;
    if (options.retriever !== undefined) agentOptions.retriever = options.retriever;
    return new Agent(agentOptions);
  }

  get agentId(): string {
    return this.manifest.id;
  }

  /** Must equal the orchestrator's agentTaskQueue() — the dispatch contract. */
  get taskQueue(): string {
    return `agent-${this.agentId}`;
  }

  /** Registers the handler for a manifest-declared capability. */
  capability(name: string, handler: Handler): void {
    const declared = this.manifest.capabilities.map((c) => c.name);
    if (!declared.includes(name)) {
      throw new Error(
        `capability ${name} is not declared in the manifest for ${this.agentId} — ` +
          `declared: ${[...declared].sort().join(', ')}. Add it to manifest.yaml first; ` +
          'the manifest is the contract.',
      );
    }
    if (this.handlers.has(name)) {
      throw new Error(`capability ${name} already has a handler`);
    }
    this.handlers.set(name, handler);
  }

  /** Every declared capability must have a handler before serving. */
  assertComplete(): void {
    const missing = this.manifest.capabilities
      .map((c) => c.name)
      .filter((name) => !this.handlers.has(name));
    if (missing.length > 0) {
      throw new Error(`manifest declares capabilities with no handler: ${missing.join(', ')}`);
    }
  }

  answerBuilder(): AnswerBuilder {
    return new AnswerBuilder();
  }

  /**
   * The execute_capability activity body: StepRequest → StepResult.
   *
   * Output is validated against the capability's declared output_schema with
   * one structured-repair retry, then the step fails typed — never a
   * best-effort parse (orchestration.md).
   */
  async execute(request: unknown): Promise<StepResult> {
    const errors = stepRequest.errors(request);
    if (errors.length > 0) {
      throw ApplicationFailure.create({
        message: `malformed StepRequest: ${errors.join('; ')}`,
        type: 'Permanent',
        nonRetryable: true,
      });
    }
    const req = request as StepRequest;
    const handler = this.handlers.get(req.capability);
    if (handler === undefined) {
      return this.failed(
        req,
        new CapabilityError(
          ErrorClass.Permanent,
          `agent ${this.agentId} has no handler for ${req.capability}`,
        ),
      );
    }

    // A contextual model (the GatewayModel) is bound to THIS step's
    // delegated identity + correlation before it reaches the handler;
    // FakeModel is not contextual, so unit tests see zero change.
    const base = this.model ?? (this.fallbackFake ??= new FakeModel());
    const bound = isContextualModel(base)
      ? base.withCallContext({
          delegatedToken: req.delegated_token,
          taskId: req.task_id,
          stepId: req.step_id,
          tenant: req.tenant,
          capability: req.capability,
        })
      : base;
    const counting = new CountingModel(bound);
    const ctx = new CapabilityContext({
      tenant: req.tenant,
      taskId: req.task_id,
      stepId: req.step_id,
      capability: req.capability,
      delegatedToken: req.delegated_token,
      budget: req.budget,
      model: counting,
      retriever: this.retriever,
      log: this.log.child({ task_id: req.task_id, step_id: req.step_id }),
    });

    const tracer = trace.getTracer('acp-agent-sdk');
    return await tracer.startActiveSpan(`invoke_agent ${req.capability}`, async (span) => {
      span.setAttributes({
        'gen_ai.operation.name': 'invoke_agent',
        'gen_ai.agent.name': this.agentId,
        'acp.tenant': req.tenant,
        'acp.task_id': req.task_id,
        'acp.capability': req.capability,
      });
      try {
        const output = await this.executeValidated(handler, ctx, req);
        span.setAttribute('acp.step_status', 'completed');
        return {
          kind: 'step_result',
          step_id: req.step_id,
          task_id: req.task_id,
          tenant: req.tenant,
          status: 'completed',
          output,
          usage: usageOf(counting),
        } satisfies StepResult;
      } catch (err) {
        if (err instanceof CapabilityError) {
          if (err.errorClass === ErrorClass.Retryable) {
            // Temporal owns retries; surface as a retryable failure.
            throw ApplicationFailure.create({
              message: err.message,
              type: 'Retryable',
              nonRetryable: false,
            });
          }
          span.setAttribute('acp.step_status', 'failed');
          return this.failed(req, err, counting);
        }
        throw err;
      } finally {
        span.end();
      }
    });
  }

  /** Serves this agent's task queue until cancelled (live Temporal + NATS; E2E covers it). */
  async run(): Promise<void> {
    const { serveAgent } = await import('./worker.js');
    await serveAgent(this);
  }

  private validatorFor(capability: string): ValidateFunction {
    const cached = this.validators.get(capability);
    if (cached !== undefined) return cached;
    const declared = this.manifest.capabilities.find((c) => c.name === capability);
    if (declared === undefined) {
      // Unreachable through execute(): handlers only register declared names.
      throw new Error(`capability ${capability} is not declared in the manifest`);
    }
    const compiled = this.ajv.compile(asRecord(declared.output_schema));
    this.validators.set(capability, compiled);
    return compiled;
  }

  private async executeValidated(
    handler: Handler,
    ctx: CapabilityContext,
    req: StepRequest,
  ): Promise<Record<string, unknown>> {
    const validate = this.validatorFor(ctx.capability);
    const input = asRecord(req.input);
    let output = await handler(ctx, input);
    let violations = violationsOf(validate, output);
    if (violations.length === 0) return output;
    ctx.log.warn(
      { errors: violations },
      'output failed schema validation; structured-repair retry',
    );
    output = await handler(ctx, input);
    violations = violationsOf(validate, output);
    if (violations.length > 0) {
      throw new CapabilityError(
        ErrorClass.Permanent,
        'handler output does not conform to the declared output_schema after one repair ' +
          `retry: ${violations.slice(0, 3).join('; ')}`,
      );
    }
    return output;
  }

  private failed(req: StepRequest, err: CapabilityError, counting?: CountingModel): StepResult {
    const result: StepResult = {
      kind: 'step_result',
      step_id: req.step_id,
      task_id: req.task_id,
      tenant: req.tenant,
      status: 'failed',
      error: err.toProtocol(),
    };
    if (counting !== undefined) result.usage = usageOf(counting);
    return result;
  }
}

function usageOf(counting: CountingModel): Usage {
  // Cache tokens and model id are emitted only when present: exactOptional
  // PropertyTypes + the protocol's additionalProperties:false mean an absent
  // field must be omitted, not sent as undefined/zero. This keeps zero-LLM
  // usage byte-identical to before the cache fields existed.
  return {
    llm_calls: counting.llmCalls,
    input_tokens: counting.inputTokens,
    output_tokens: counting.outputTokens,
    ...(counting.cacheReadTokens > 0 ? { cache_read_tokens: counting.cacheReadTokens } : {}),
    ...(counting.cacheWriteTokens > 0 ? { cache_write_tokens: counting.cacheWriteTokens } : {}),
    ...(counting.model !== undefined ? { model: counting.model } : {}),
  };
}

function violationsOf(validate: ValidateFunction, output: unknown): string[] {
  if (validate(output)) return [];
  return (validate.errors ?? []).map((e) => `${e.instancePath || '/'} ${e.message ?? 'invalid'}`);
}
